# Summary

Created on Thu May 7 2020 <br>
@author: goutham <br>

## Multi Label Classifcation using keras Deep Neural Networks <br>

Topics
* Keras Neural Networks
* Multi Label Classification 
* Metric = 'accuracy'
* Model Evaluation using Custom Metric <br>

Packages Used: 
* keras
* numpy
* pandas <br>

The purpose of this code is to show how one can build a basic Multi Label Classification Neural Network using Keras.
This code also helps one understand what to look out for when using 'accuracy' as a metric for model evaluation in multi label classification problems with high target dimensionality and sparse data.



## Get the Data

In [0]:
# Input the Data
import pandas as pd
import numpy as np
dataGroup = pd.read_csv('YourPath/MLCSampleData.csv') # Insert your data path here

In [3]:
# Understand the Data
dataGroup.head()
# This is dummy data that I created to use for building a multi label classification keras neural network
# The label data is one-hot encoded - the way it needs to be fed into the output layer
# id column is a unique identifier for an instance 
# Feature0 to Feature99 are the 100 features that should be used to build the classifier
# Label0 to Label39 are the 40 distinct labels (output) available and the data is sparse

id  Feature0  Feature1  Feature2  ...  Label36  Label37  Label38  Label39
0   1  0.015835 -0.045127  0.094640  ...        0        0        0        0
1   2 -0.063157 -0.017947 -0.030838  ...        0        0        0        0
2   3 -0.037002 -0.030174  0.080122  ...        0        0        0        1
3   4 -0.041543 -0.020182  0.002726  ...        0        0        0        0
4   5 -0.041543 -0.020182  0.002726  ...        0        0        0        0

[5 rows x 141 columns]

In [0]:
# Train Test Split : Random Split 90:10 # Usually one does 80:20
X = dataGroup.iloc[:,1:101].copy() # get Feature0 to Feature99
y = dataGroup.iloc[:,101:].copy() # get Label0 to Label39
from sklearn.model_selection import train_test_split # random sampling split
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.10, random_state=0) # using random_state=0 for replication of results # should not use it when deploying

# transform the (Y) target into a numy array
y_train_list = y_train.values.tolist()
y_train = np.array(y_train_list)

y_test_list = y_test.values.tolist()
y_test = np.array(y_test_list)

## Build the Classifier

A very simple neural network was built as the primary objective was to focus on how the output layer and the loss function differs in a multi label classification from that of a conventional multi class classification network. <br>
This could be used as a base to build complex multi label classifiers.

In [5]:
# use keras framework for deep learning
from tensorflow import keras
from keras.models import Sequential
from keras.layers import Dense

Using TensorFlow backend.


In [6]:
# load a sequential model
model = Sequential()

# construct the network
model.add(Dense(units=64, activation='relu', input_dim=len(x_train.columns))) # len(x_train.columns) will automatically get the required count of nodes for input layer
model.add(Dense(units=128, activation='relu'))
model.add(Dense(len(y_train[0]), activation='sigmoid')) # Sigmoid over Softmax although we have multiple classes for independence across labels
                                                        # len(y_train[0]) = number of labels = number of nodes in output layer

# compile the model # using binary crossentropy over categorical crossentropy for independence across labels
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy']) 
# I'm using metrics=['accuracy']  in model.compile just to show what happens in model evaluation step later
# If the number of distinct labels is high and the data is sparse, i would ideally compile the model using the below line
#model.compile(loss='binary_crossentropy', optimizer='adam') 

# y_train is a Numpy arrays 
model.fit(x_train, y_train, epochs=150, batch_size=36, verbose = 0) # Using verbose = 0 as the interim output will make the notebook huge

Let's see what the model evaluation tells us

In [7]:
# Evaluate the model on the test data using 'evaluate'
print("Output when model evaluation was performed using model.evaluate() and ")
print("model compiled using <model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])>")
print()
results = model.evaluate(x_test, y_test, verbose = 0)
print("Test Accuracy","{:.2%}".format(results[1]))

Output when model evaluation was performed using model.evaluate() and 
model compiled using <model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])>

Test Accuracy 97.22%


Let's see how many records were accurately labelled by the model - our own evaluation metric

In [8]:
# Predict on test data using the model
labelsPredict = model.predict(x_test)

# I'm going to call a label 1 if its value > 0.5 and so, im going to use round function to get labels
# Threshold could be set dynamically for better performance of the model
labelsPredict = labelsPredict.round()

# Create Multi Label Vectors for the observed data # look at the output below to understand what this transformation does 
observed = []
for i in range(len(y_test)):
  multiLabel = ''
  for j in range(len(y_test[i,:])):
    multiLabel = multiLabel+str(y_test[i,j])
  observed.append(multiLabel)
print('Observed Data - first 5 records')
print(observed[:5])
print()

# Create Multi Label Vectors for the data predicted using model 1 # look at the output below to understand what this transformation does 
predicted = []
for i in range(len(labelsPredict)):
  multiLabel = ''
  for j in range(len(labelsPredict[i])):
    multiLabel = multiLabel+str(int(labelsPredict[i,j]))
  predicted.append(multiLabel)
print('Predicted Data - first 5 records')
print(predicted[:5])
print()

# For how many records did the model get all the labels correct?
is_accurate = []
for i in range(len(observed)):
  is_accurate.append(observed[i]==predicted[i])

CategoricalAccuracy = sum(is_accurate)/len(is_accurate) # (count of records accurately tagged)/(total number of records)
print('% of test records for which the model got all the labels correct is')
print("{:.2%}".format(CategoricalAccuracy))  
print()
print("total number of test records:",len(is_accurate))
print("accurately predicted records:",sum(is_accurate))


Observed Data - first 5 records
['0000000000000000000000100000000000000000', '0000000001000000000000000000000000000000', '0000010000000000000000000000000000000000', '0000000000000000000000000000000000000100', '0000000000000000000000100000000000000000']

Predicted Data - first 5 records
['0000000000000100000000000000000000010000', '0000000000000000000000000000000000000000', '0000000000000000000000000000000000000000', '0000000000000000000001000000000000000000', '0000000000000000000000100000000000000000']

% of test records for which the model got all the labels correct is
30.16%

total number of test records: 63
accurately predicted records: 19


As we can see, the code in the above cell calls the predictions for a record accurate only if the model gets all the labels correct for that record. But, the keras default accuracy metric considers each prediction as an individual event and so the discrepancy.

# Conclusion
There is **nothing wrong** with the keras accuracy metric. <br>
It is working the way it is intended to work. It considers each record-label prediction as an individual event.<br>
I've written a story on Medium explaining how 'Accuracy' works for a multi label classification problem and how to avoid any pitfalls that may arise from misinterpreting it.
Please refer to it for a non technical explanation on why the model evaluation is giving a test accuracy of 97% when the model only got multi-labels correct for 30% of the data.